File: tools/video-marketing/scripts/transcribe.mjs
#!/usr/bin/env node
/**
 * Transcripción en español con tiempos por palabra.
 *
 * Por defecto usa faster-whisper EN LOCAL: gratis, sin límite y sin mandar
 * tu grabación a ningún sitio. Con --fal usa Whisper en fal.ai, que sale a
 * céntimos, por si no quieres instalar Python.
 *
 * USO:
 *   node scripts/transcribe.mjs grabacion.mp4 transcript.json
 *   node scripts/transcribe.mjs grabacion.mp4 transcript.json --model medium
 */
import './env.mjs'
import { execFileSync } from 'node:child_process'
import { existsSync, writeFileSync } from 'node:fs'
import { dirname, join } from 'node:path'
import { fileURLToPath } from 'node:url'
import { falRun } from './fal.mjs'

const HERE = dirname(fileURLToPath(import.meta.url))

export async function transcribe(input, output, { model = 'small', useFal = false } = {}) {
  if (!existsSync(input)) throw new Error(`No existe el archivo: ${input}`)

  if (useFal) {
    const res = await falRun('fal-ai/whisper', {
      audio_url: input,        // fal acepta URL pública o data URI
      task: 'transcribe',
      language: 'es',
      chunk_level: 'word',
    })
    writeFileSync(output, JSON.stringify(res, null, 2))
    return output
  }

  try {
    execFileSync('python3', [join(HERE, 'py', 'transcribe.py'), input, output, model],
      { stdio: 'inherit' })
  } catch (err) {
    throw new Error(
      'La transcripción local falló. Comprueba que tienes Python y faster-whisper:\n' +
      '  pip install faster-whisper\n' +
      'O usa --fal para transcribir en la nube.'
    )
  }
  return output
}

// Ejecutable directo
if (import.meta.url === `file://${process.argv[1]}`) {
  const [input, output] = process.argv.slice(2).filter(a => !a.startsWith('--'))
  const argv  = process.argv.slice(2)
  const model = argv.includes('--model') ? argv[argv.indexOf('--model') + 1] : 'small'
  if (!input || !output) {
    console.error('uso: node scripts/transcribe.mjs <vídeo> <salida.json> [--model small|medium] [--fal]')
    process.exit(2)
  }
  await transcribe(input, output, { model, useFal: argv.includes('--fal') })
}

File: tools/video-marketing/scripts/py/transcribe.py
#!/usr/bin/env python3
"""
Transcribe un audio/vídeo al español con faster-whisper y escupe JSON con
tiempos por palabra.

Corre en local y es gratis: el modelo se descarga una vez (~150 MB el
'small', que ya va sobrado para voz limpia grabada a propósito).

  python3 transcribe.py entrada.mp4 salida.json [modelo]
"""
import json
import sys


def main() -> int:
    if len(sys.argv) < 3:
        print("uso: transcribe.py <entrada> <salida.json> [modelo]", file=sys.stderr)
        return 2

    src, dst = sys.argv[1], sys.argv[2]
    model_name = sys.argv[3] if len(sys.argv) > 3 else "small"

    try:
        from faster_whisper import WhisperModel
    except ImportError:
        print("Falta faster-whisper. Instálalo con:  pip install faster-whisper",
              file=sys.stderr)
        return 1

    model = WhisperModel(model_name, device="cpu", compute_type="int8")
    segments, info = model.transcribe(
        src,
        language="es",
        word_timestamps=True,
        vad_filter=True,          # ignora los silencios, menos alucinación
    )

    out = []
    for seg in segments:
        out.append({
            "start": round(seg.start, 3),
            "end":   round(seg.end, 3),
            "text":  seg.text.strip(),
            "words": [
                {"start": round(w.start, 3), "end": round(w.end, 3), "word": w.word.strip()}
                for w in (seg.words or [])
            ],
        })
        print(f"  [{seg.start:6.2f}] {seg.text.strip()[:70]}", file=sys.stderr)

    with open(dst, "w", encoding="utf-8") as f:
        json.dump({"language": info.language, "duration": info.duration, "segments": out},
                  f, ensure_ascii=False, indent=2)

    print(f"✓ {len(out)} segmentos → {dst}", file=sys.stderr)
    return 0


if __name__ == "__main__":
    sys.exit(main())
